### ISSUE TITLE
ServiceNow connector client (8.9/8.10 Beta)

### ISSUE THREAD

**Issue by danajuratoni** [elastic/connectors-python#545]
## New ServiceNow connector
New connector client to be released with 8.9 in Beta, more performant version in 8.10.
Preliminary date for 8.9 Feature Freeze: Tue, Jun 20 2023

### Acceptance criteria
- [x] TDD in [GDrive](https://drive.google.com/drive/folders/1DN2y4TRXIuybX8vawHljkWozoyPEOYaf)
- [x] TDD reviewed by engineer and PM
- [x] QA plan in [GDrive](https://drive.google.com/drive/folders/1DN2y4TRXIuybX8vawHljkWozoyPEOYaf)
- [x] QA plan reviewed by engineer and PM
- [x] Code merged in main following the defined PR process
     - [x] ServiceNow: Ping feature https://github.com/elastic/connectors-python/pull/815
     - [x] #877
     - [x] #880
     - [x] #1033
- [x] https://github.com/elastic/connectors-python/issues/948
- [x] Functional tests performed
- [x] Performance test/ reports against the last stable release branch of the connector framework and in Elastic Cloud 
- [x] Connector client documentation merged in repo 
    - [x] #898
- [x] Demo of the connector in latest builds in Elastic Cloud in [GDrive](https://drive.google.com/drive/folders/1DN2y4TRXIuybX8vawHljkWozoyPEOYaf)
- [x] Sign off given from an engineer in Ingestion
     - [x] Code backported to the correct version 
     - [x] Automated tests to 90% coverage
- [x] Regression test suite performed for the release along with QA day for all supported OS in https://www.elastic.co/support/matrix 
- [x] Compatibility tests for running connector as native in Cloud

**Comment by danajuratoni** [comment]
@prashant-elastic could you please assign an engineer as discussed?
@timgrein updated the AC with docs "reviewed by engineer and PM", marking the TDD as reviewed based on your comment in chat.

**Comment by parth-elastic** [comment]
Sharing last week's progress summary (4th May - 10th May): 

- Resolve comments and Merged following PRs
   - #876 
   - #881
- Working on documentation

**Comment by parth-elastic** [comment]
Progress summary [11th May - 17th May]:

Raised PR for documentation

- https://github.com/elastic/connectors-python/pull/899

Completed performance test, generated report and shared in the elastic folder - [Click Here For the Doc](https://docs.google.com/document/d/1guB3BECH_lXBHztx7jpCRZG9YYfuQcRGf9ZILt3TCFw/edit?usp=share_link)

**Comment by parth-elastic** [comment]
Progress summary [18th May - 25th May]:

Addressed comments on the documentation PR

https://github.com/elastic/connectors-python/pull/899

Addressed comments on performance test - [Click Here For the Doc](https://docs.google.com/document/d/1guB3BECH_lXBHztx7jpCRZG9YYfuQcRGf9ZILt3TCFw/edit?usp=share_link)

Awaiting re-review for both PR & Performance report.

**Comment by parth-elastic** [comment]
Progress summary [26th May - 8th June]:

- Started working on adding a support for batching #1033 
- ServiceNow Documentation PR is Merged #899 . Closed the issue for the same too.

**Comment by parth-elastic** [comment]
Progress summary [9th June - 15th June]:


- Raised a PR for batching https://github.com/elastic/connectors-python/issues/1033
- Updated E2E as per the batching in the same PR.
- Performance test is in progress, Will be sharing the stats & perf8 report soon.

**Comment by navarone-feekery** [comment]
@parth-elastic how is this connector going? I understand that the base of it has been merged into main. Is there anything left that would prevent us from landing this in 8.9?

**Comment by parth-elastic** [comment]
Progress summary [16th June - 22th June]:

- All the Functional PRs are merged into main.
- Started working on Advance Sync rules, Added the test cases of sync rules [here in the GDrive](https://docs.google.com/spreadsheets/d/1Wn7Ajt-1t6Y5C-LA91DITD2R6bf15cPA/edit#gid=788072971) 
- Facing `MEMQueue` error in performance testing, trying alternatives for the same.

**Comment by parth-elastic** [comment]
Progress summary [23th June - 29th June]:


- Merged the PR for Advance Sync rules
- Merged the PR for Documentation for Advance Sync Rules
- Trying different batch/fetch values to generate Performance report. @khushbu-elastic will share the doc soon.

**Comment by khushbu-elastic** [comment]
@danajuratoni I have added the performance stats in the shared google drive folder for servicenow and shared the link in the slack.

**Comment by danajuratoni** [comment]
@khushbu-elastic 
Looking at the performance [results without batching](https://docs.google.com/document/d/1guB3BECH_lXBHztx7jpCRZG9YYfuQcRGf9ZILt3TCFw/edit#bookmark=id.44jgf2ewzmrj) and [with batching](https://docs.google.com/document/d/1dYxW1PK6zSFfLhD8VPSde5lHTgkumV-CWgsxsybImSw/edit#bookmark=id.afakuxnp60t9) they look almost identical. The expected performance improvement was around 30% - am I misreading anything?

**Comment by parth-elastic** [comment]
@danajuratoni Khushbu is on PTO for this week. Let me answer on her behalf.

> Looking at the performance [results without batching](https://docs.google.com/document/d/1guB3BECH_lXBHztx7jpCRZG9YYfuQcRGf9ZILt3TCFw/edit#bookmark=id.44jgf2ewzmrj) and [with batching](https://docs.google.com/document/d/1dYxW1PK6zSFfLhD8VPSde5lHTgkumV-CWgsxsybImSw/edit#bookmark=id.afakuxnp60t9) they look almost identical. The expected performance improvement was around 30% - am I misreading anything?

Stats are looking similar because, Based on our discussion with @seanstory on [slack](https://crestdatasystems.slack.com/archives/C02AV5C8ZDE/p1687169816695739), we have reduced batch/fetch size to successfully run the connector as we were facing MEMQUUE error. once the disc based queue is implemented we can make ServiceNow faster 

> What is the progress with the last remaining ACs?

We are waiting on the latest BC of 8.9 to perform the remaining ACs. All the remaining ACs needs to be tested on the latest BC. So once that is available @khushbu-elastic will pick those ACs.

**Comment by khushbu-elastic** [comment]
@danajuratoni @DianaJourdan 
We have verified the below two points so that you can mark as done in this task.:

- Demo of the connector in latest builds in Elastic Cloud in [Uploaded to GDrive] 

- Regression test suite performed for the release along with QA day for all supported OS (CentOS7, Alma Linux8, Rocky Linux8, Mac OS)

 I was not able to check the servicenow connector as a Native in cloud as the 8.9 SNAPSHOT as the Sync is not getting started and shows the Status as Sync Pending
- Compatibility tests for running connector as native in Cloud

**Comment by khushbu-elastic** [comment]
@danajuratoni @DianaJourdan I have checked the servicenow connector as a Native in cloud as the 8.9 SNAPSHOT as we discussed in slack so you can mark the below point as done and close this issue after verification.

- Compatibility tests for running connector as native in Cloud

Uploaded the demo here for the same - [ServiceNow as Native Connector](https://drive.google.com/file/d/1EJJNni9VrGSqKh0QoHjC46tu-_dpueJ6/view?usp=drive_link)

**Comment by danajuratoni** [comment]
Nice work! Closing this as done 🎉
